RakeshRenzous/space-rockets
# perf: Virtualizing the launches page grid

**Description**

In the list page of launches we have a lot of content that is being rendered as we keep on scrolling by clicking on the "Show more" button, in this we are rendering a lot of content on the page and this overall increases the page weight. Hence I thought we can virtualise the entire grid to improve the performance of the page. 

**Thoughts**

Started by exploring react-window and react-virtualised but neither have a proper support for grid based layouts while react-virtuso was more versatile when compared to the former. 

Have also moved to Infinite scrolling instead of having load more button has this encourages discoverability of the content. As of now did not refactor the launch-pads as the content is currently to small to leverage this enhancement. So in the future when there comes time if we need to introduce this launch-pads or even a new page we can simply reuse the ```VirtualizedGrid``` component.

**Steps to test**

- Visit home page and click on "Browse SpaceX Launches".
- Scroll to the bottom and you should see more content automatically loaded.

To see it working

- Right click on the web page and click "Inspect" in the context menu.
- When scrolling, In the DOM tree you can see only the content that is visible in the viewport and some extra content would be rendered instead of the entire grid.

**_Both images after loading all the content_** 

**Before**

<img width="381" alt="Screenshot 2021-11-22 at 6 13 45 AM" src="https://user-images.githubusercontent.com/23741529/142785990-917326b4-6411-4ca0-a67f-2010874a1970.png">


**After**

<img width="380" alt="Screenshot 2021-11-22 at 6 14 18 AM" src="https://user-images.githubusercontent.com/23741529/142785957-d52f0660-5378-4ffc-93d6-ff8ed8b25d9d.png">
